File: info/2. Структуры данных/1. Cписки, стеки, очереди, ассоциативные массивы/9. Очередь на основе массива/code sample/Java/Queue.java


public class Queue {
    
    private final int DEFAULT_SIZE = 16;
    private Object[] data;
    private int head;
    private int tail;
    private int size;

    public Queue() {
        super();
        data = new Object[DEFAULT_SIZE];
        head = 0;
        tail = 0;
        size = 0;
    }

    public void enqueue(Object value) {
        if ((tail + 1) % data.length == head) {
            increaseSize();
        }
        data[tail] = value;
        tail = (tail + 1) % data.length;
        size += 1;
    }

    private void increaseSize() {
        if (data.length >= Integer.MAX_VALUE - 10) {
            throw new IllegalArgumentException("can't increase the size");
        }
        Object[] newArray = new Object[Math.min(Integer.MAX_VALUE - 10, data.length * 4 / 3 + 1)];
        int addIndex = 0;
        for (;;) {
            if (head % data.length == tail) {
                break;
            }
            newArray[addIndex] = data[head % data.length];
            addIndex += 1;
            head = (head + 1) % data.length;
        }
        data = newArray;
        head = 0;
        tail = addIndex;
        size = tail;
    }

    public Object dequeue() {
        if (head == tail) {
            return null;
        }
        Object returnValue = data[head];
        data[head] = null;
        head = (head + 1) % data.length;
        size -= 1;
        return returnValue;
    }

    public Object peek() {
        if (head == tail) {
            return null;
        }
        return data[head];
    }

    public void trimToSize() {
        Object[] newArray = new Object[size + 1];
        int addIndex = 0;
        for (;;) {
            if (head % data.length == tail) {
                break;
            }
            newArray[addIndex] = data[head % data.length];
            addIndex += 1;
            head = (head + 1) % data.length;
        }
        data = newArray;
        head = 0;
        tail = addIndex;
    }

    public String toString() {
        String result = "[";
        int currentIndex = (tail - 1) >= 0 ? tail - 1 : data.length - 1;
        for (int i = size; i > 0; i--) {
            result += data[currentIndex] + ", ";
            currentIndex = (currentIndex - 1) >= 0 ? currentIndex - 1 : data.length - 1;
        }
        return result.substring(0, result.length() - 2) + "]";
    }

}

File: info/2. Структуры данных/1. Cписки, стеки, очереди, ассоциативные массивы/9. Очередь на основе массива/code sample/Fortran/queue.f90
module queue

implicit none


type Array_Queue
    integer,allocatable::data_array(:)
    integer::tail, head, queue_size
    
    contains
        procedure,pass::init
        procedure,pass::enqueue
        procedure,pass::up_resize
        procedure,pass::dequeue
        procedure,pass::peek
        procedure,pass::trim_to_size
        procedure,pass::clear
        procedure,pass::destroy
        procedure,pass::show
end type Array_Queue
        
contains

    subroutine init(this)
        class(Array_Queue)::this
        if(.not. allocated(this%data_array)) then
            allocate(this%data_array(16))
            this%tail = 1
            this%head = 1
            this%queue_size = 0
        end if
    end subroutine init
    
    subroutine up_resize(this)
        class(Array_Queue)::this
        integer,allocatable::temp_array(:)
        integer::add_index, new_size, old_size
        old_size = size(this%data_array)
        new_size = old_size * 4/3 + 1
        allocate(temp_array(new_size))
        add_index = 1
        do
            if(this%head == this%tail) then
                exit
            end if
            temp_array(add_index) = this%data_array(this%head)
            add_index = add_index + 1
            this%head = this%head + 1
            if(this%head > size(this%data_array)) then
                this%head = 1
            end if
        end do
        this%data_array = temp_array
        this%head = 1
        this%tail = add_index
        this%queue_size = add_index - 1
        deallocate(temp_array)
    end subroutine up_resize
         

    subroutine enqueue(this, data_value)
        class(Array_Queue)::this
        integer, intent(in)::data_value
        integer:: current_size, next_index
        current_size = size(this%data_array)
        next_index = this%tail + 1
        if (next_index > current_size) then
            next_index = 1
        end if
        if (next_index == this%head) then
            call this%up_resize()
        end if
        this%data_array(this%tail) = data_value
        this%tail = this%tail + 1
        if (this%tail > size(this%data_array)) then
            this%tail = 1
        end if
        this%queue_size = this%queue_size + 1
    end subroutine enqueue
    
    subroutine dequeue(this, element, op_result)
        class(Array_Queue)::this
        integer, intent(inout)::element
        logical, intent(inout)::op_result
        if(this%head == this%tail) then
            op_result = .false.
            return
        end if
        element = this%data_array(this%head)
        op_result = .true.
        this%head = this%head + 1
        if(this%head > size(this%data_array)) then
            this%head = 1
        end if
        this%queue_size = this%queue_size - 1
    end subroutine dequeue
    
    subroutine peek(this, element, op_result)
        class(Array_Queue)::this
        integer, intent(inout)::element
        logical, intent(inout)::op_result
        if(this%head == this%tail) then
            op_result = .false.
            return
        end if
        element = this%data_array(this%head)
        op_result = .true.
    end subroutine peek
    
    subroutine trim_to_size(this)
        class(Array_Queue)::this
        integer::add_index
        integer, allocatable::new_array(:)
        allocate(new_array(this%queue_size + 1))
        add_index = 1
        do
            if(this%head == this%tail) then
                exit
            end if
            new_array(add_index) = this%data_array(this%head)
            add_index = add_index + 1
            this%head = this%head + 1
            if(this%head > size(this%data_array)) then
                this%head = 1
            end if
        end do
        this%data_array = new_array
        this%head = 1
        this%tail = add_index
        this%queue_size = add_index - 1
        deallocate(new_array)
    end subroutine trim_to_size
            
    subroutine clear(this)
        class(Array_Queue)::this
        if(allocated(this%data_array)) then
            deallocate(this%data_array)
            allocate(this%data_array(16))
        end if
        this%head = 1
        this%tail = 1
        this%queue_size = 0
    end subroutine clear
    
    subroutine destroy(this)
        class(Array_Queue)::this
        if(allocated(this%data_array)) then
            deallocate(this%data_array)
        end if
    end subroutine destroy
    
    subroutine show(this)
        class (Array_Queue)::this
        integer::i, j
        i = this%tail - 1
        if (i < 1) then
            i = size(this%data_array)
        end if
        write(*,'(A1)',advance = 'NO') '['
        do j = 1, this%queue_size
            write(*,'(I5)',advance = 'NO') this%data_array(i)
            i = i - 1
            if (i < 1) then
                i = size(this%data_array)
            end if
        end do
        write(*,*)']'
    end subroutine show
            
end module queue
